### PR TITLE
rgw/multisite: add cls versioning for tracking data sync per shard object

### PR DESCRIPTION
**Pull request by smanjara**
Replaces #47614

Also stores the RGWObjVersionTracker's in a vector. 

Signed-off-by: Shilpa Jagannath <smanjara@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_data_sync.cc [hunk 1]
```diff
@@ -981,13 +997,15 @@ class RGWListBucketIndexesCR : public RGWCoroutine {
...
+          objv.clear();
           spawn(new RGWSimpleRadosWriteCR<rgw_data_sync_marker>(
 		  dpp, sync_env->async_rados, sync_env->svc->sysobj,
 		  rgw_raw_obj(sync_env->svc->zone->get_zone_params().log_pool,
 			      RGWDataSyncStatusManager::shard_obj_name(
 				sc->source_zone, shard_id)),
-		  marker),
+		  marker, &objv),
 		true);
+          objvs[shard_id] = objv;
```
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc:1008]: we can pass `&objvs[shard_id]` directly to `RGWSimpleRadosWriteCR` here. and we shouldn't call `clear()` before this write - we need to use the same objv from our earlier call to `RGWReadDataSyncStatusCoroutine` or `RGWInitDataSyncStatusCoroutine`

File: src/rgw/rgw_data_sync.cc [hunk 2]
```diff
@@ -2231,7 +2260,7 @@ class RGWDataSyncCR : public RGWCoroutine {
           for (map<uint32_t, rgw_data_sync_marker>::iterator iter = sync_status.sync_markers.begin();
                iter != sync_status.sync_markers.end(); ++iter) {
             RGWDataSyncShardControlCR *cr = new RGWDataSyncShardControlCR(sc, sync_env->svc->zone->get_zone_params().log_pool,
-                                                                          iter->first, iter->second, sync_status, tn);
+                                                                          iter->first, iter->second, sync_status, objvs[iter->first], tn);
```
Comment (cbodley) [inline src/rgw/rgw_data_sync.cc:2263]: this and everything under `RGWDataSyncShardControlCR` look great 👍

File: src/rgw/rgw_data_sync.cc [hunk 3]
```diff
@@ -127,11 +130,13 @@ bool RGWReadDataSyncStatusMarkersCR::spawn_next()
     return false;
   }
   using CR = RGWSimpleRadosReadCR<rgw_data_sync_marker>;
+  objv.clear();
   spawn(new CR(env->dpp, env->async_rados, env->svc->sysobj,
                rgw_raw_obj(env->svc->zone->get_zone_params().log_pool, RGWDataSyncStatusManager::shard_obj_name(sc->source_zone, shard_id)),
-               &markers[shard_id]),
+               &markers[shard_id], &objv),
```
[the remark this answers is not in the document]
Comment (smanjara) [reply src/rgw/rgw_data_sync.cc]: fixed

File: src/rgw/rgw_data_sync.cc [hunk 4]
```diff
@@ -214,7 +221,7 @@ int RGWReadDataSyncStatusCoroutine::operate(const DoutPrefixProvider *dpp)
     // read shard markers
     using ReadMarkersCR = RGWReadDataSyncStatusMarkersCR;
     yield call(new ReadMarkersCR(sc, sync_status->sync_info.num_shards,
-                                 sync_status->sync_markers));
+                                 sync_status->sync_markers, objvs));
```
[the remark this answers is not in the document]
Comment (smanjara) [reply src/rgw/rgw_data_sync.cc:222]: fixed

File: src/rgw/rgw_data_sync.cc [hunk 5]
```diff
@@ -621,9 +631,11 @@ class RGWInitDataSyncStatusCoroutine : public RGWCoroutine {
           marker.next_step_marker = info.marker;
           marker.timestamp = info.last_update;
           const auto& oid = RGWDataSyncStatusManager::shard_obj_name(sc->source_zone, i);
+          objv.generate_new_write_ver(cct);
           using WriteMarkerCR = RGWSimpleRadosWriteCR<rgw_data_sync_marker>;
           spawn(new WriteMarkerCR(dpp, sync_env->async_rados, sync_env->svc->sysobj,
-                                  rgw_raw_obj{pool, oid}, marker), true);
+                                  rgw_raw_obj{pool, oid}, marker, &objv), true);
+          objvs.push_back(objv);
```
[the remark this answers is not in the document]
Comment (smanjara) [reply src/rgw/rgw_data_sync.cc]: fixed

File: src/rgw/rgw_data_sync.cc (hunk 1 above)
Comment (smanjara) [reply src/rgw/rgw_data_sync.cc]: fixed

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
from make check:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_data_sync.cc
/home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_data_sync.cc:134:37: error: reference cannot be bound to dereferenced null pointer in well-defined C++ code; pointer may be assumed to always convert to true [-Werror,-Wundefined-bool-conversion]
               &markers[shard_id], &objvs[shard_id]),
                                    ^~~~~~~~~~~~~~~
```

**Comment by smanjara** [comment]
> from make check:
> 
> ```
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_data_sync.cc
> /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_data_sync.cc:134:37: error: reference cannot be bound to dereferenced null pointer in well-defined C++ code; pointer may be assumed to always convert to true [-Werror,-Wundefined-bool-conversion]
>                &markers[shard_id], &objvs[shard_id]),
>                                     ^~~~~~~~~~~~~~~
> ```

oh, I didn't see this error when I built locally.

**Comment by smanjara** [comment]
jenkins make check

**Comment by smanjara** [comment]
> > from make check:
> > ```
> > /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_data_sync.cc
> > /home/jenkins-build/build/workspace/ceph-pull-requests/src/rgw/rgw_data_sync.cc:134:37: error: reference cannot be bound to dereferenced null pointer in well-defined C++ code; pointer may be assumed to always convert to true [-Werror,-Wundefined-bool-conversion]
> >                &markers[shard_id], &objvs[shard_id]),
> >                                     ^~~~~~~~~~~~~~~
> > ```
> 
> oh, I didn't see this error when I built locally.

thanks to @adamemerson! Pushed an update

**Comment by smanjara** [comment]
jenkins make check

**Comment by smanjara** [comment]
unrelated make check failure. @cbodley should I run tests on this pr separately or can we batch it up with other pr's?

**Comment by cbodley** [comment]
> unrelated make check failure. @cbodley should I run tests on this pr separately or can we batch it up with other pr's?

batching is fine 👍

**Comment by smanjara** [comment]
some multisite test failures
http://pulpito.front.sepia.ceph.com/smanjara-2022-09-22_02:46:16-rgw:multisite-rgw-cls-version-sync-shard-distro-default-smithi/7040862/

**Comment by cbodley** [comment]
yeah, there are several segfaults in there. can you try rebasing to make sure we have all the latest fixes?

**Comment by smanjara** [comment]
> yeah, there are several segfaults in there. can you try rebasing to make sure we have all the latest fixes?

ran the suite again after rebasing.

http://qa-proxy.ceph.com/teuthology/smanjara-2022-10-06_13:48:47-rgw:multisite-rgw-cls-version-sync-shard-distro-default-smithi/7057068/teuthology.log

we seem to have sync failures

**Comment by cbodley** [comment]
this will merge through https://github.com/ceph/ceph/pull/48898

**Pull request closed without merging**